server/src/services/organizations.service.ts

```diff
@@ -539,4 +551,101 @@ export class OrganizationsService {
     }
     return organizationUser;
   }
+
+  decamelizeDefaultGroupNames(groups) {
+    return groups
+      .split('|')
+      .map((group: string) =>
+        group === 'All Users' || group === 'Admin' ? decamelize(group.replace(' ', '')) : group
+      );
+  }
+
+  async bulkUploadUsers(currentUser: User, fileStream, res: Response) {
+    const users = [];
+    const existingUsers = [];
+    const invalidRows = [];
+    const invalidGroups = [];
+    const emailPattern = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i;
+
+    csv
+      .parseString(fileStream.toString(), {
+        headers: ['first_name', 'last_name', 'email', 'groups'],
+        renameHeaders: true,
+        ignoreEmpty: true,
+      })
+      .transform((row: UserCsvRow, next) => {
+        return next(null, {
+          ...row,
+          groups: this.decamelizeDefaultGroupNames(row?.groups),
+        });
+      })
+      .validate(async (data: UserCsvRow, next) => {
+        await dbTransactionWrap(async (manager: EntityManager) => {
+          //Check for existing users
+          const user = await this.usersService.findByEmail(data?.email);
+          if (user?.organizationUsers?.some((ou) => ou.organizationId === currentUser.organizationId)) {
+            existingUsers.push(data?.email);
+          } else {
+            users.push(data);
+          }
+
+          //Check for invalid groups
+          const receivedGroups = data?.groups;
+          for (const group of receivedGroups) {
+            const orgGroupPermission = await manager.findOne(GroupPermission, {
+              where: {
+                organizationId: currentUser.organizationId,
+                group: group,
+              },
+            });
+
+            if (!orgGroupPermission) {
+              invalidGroups.push(group);
+            }
+          }
+          return next(null, data.first_name !== '' && data.last_name !== '' && emailPattern.test(data.email));
+        });
+      })
+      .on('data', function () {})
+      .on('data-invalid', (row, rowNumber) => {
+        invalidRows.push(rowNumber);
+      })
+      .on('end', async (rowCount: number) => {
+        try {
+          if (rowCount > MAX_ROW_COUNT) {
```

Comment:
Best If we can do this check before validations, but I don't think its possible with fast-csv. @vjaris42 wdyt?

Reply:
Actually no we can't do that and that's just not the limitation with fast-csv but any other library as well because the data is being processed in chunks so once it goes through the data only then we can check the row count.